libbeat/outputs/shipper/shipper.go

```diff
@@ -0,0 +1,221 @@
+	beat beat.Info,
+	observer outputs.Observer,
+	cfg *common.Config,
+) (outputs.Group, error) {
+
+	config := defaultConfig()
+	err := cfg.Unpack(&config)
+	if err != nil {
+		return outputs.Fail(err)
+	}
+
+	tls, err := tlscommon.LoadTLSConfig(config.TLS)
+	if err != nil {
+		return outputs.Fail(fmt.Errorf("invalid shipper TLS configuration: %w", err))
+	}
+
+	var creds credentials.TransportCredentials
+	if config.TLS != nil && config.TLS.Enabled != nil && *config.TLS.Enabled {
+		creds = credentials.NewTLS(tls.ToConfig())
+	} else {
+		creds = insecure.NewCredentials()
+	}
+
+	opts := []grpc.DialOption{
+		grpc.WithConnectParams(grpc.ConnectParams{
+			Backoff: backoff.Config(config.Backoff),
+		}),
+		grpc.WithTransportCredentials(creds),
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), config.Timeout)
+	defer cancel()
+
+	log := logp.NewLogger("shipper")
+	log.Debugf("trying to connect to %s...", config.Server)
+
+	conn, err := grpc.DialContext(ctx, config.Server, opts...)
+	if err != nil {
+		return outputs.Fail(fmt.Errorf("shipper connection failed with: %w", err))
+	}
+	log.Debugf("connect to %s established.", config.Server)
+
+	s := &shipper{
+		log:      log,
+		observer: observer,
+		conn:     conn,
+		client:   sc.NewProducerClient(conn),
+		timeout:  config.Timeout,
+	}
+
+	return outputs.Success(config.BulkMaxSize, config.MaxRetries, s)
+}
+
+//nolint: godox // this implementation is not finished
+func (c *shipper) Publish(ctx context.Context, batch publisher.Batch) error {
+	st := c.observer
+	events := batch.Events()
+	st.NewBatch(len(events))
+
+	dropped := 0
+
+	grpcEvents := make([]*sc.Event, 0, len(events))
+
+	c.log.Debugf("converting %d events to protobuf...", len(events))
+
+	for i, e := range events {
+
+		meta, err := convertMapStr(e.Content.Meta)
+		if err != nil {
+			// conversion errors are not recoverable, so we have to drop the event completely
+			c.log.Errorf("%d/%d failed to convert event metadata to protobuf: %w", i+1, len(events), err)
+			dropped++
+			continue
+		}
+
+		fields, err := convertMapStr(e.Content.Fields)
+		if err != nil {
+			c.log.Errorf("%d/%d failed to convert event fields to protobuf: %w", i+1, len(events), err)
+			dropped++
+			continue
+		}
+
+		grpcEvents = append(grpcEvents, &sc.Event{
+			Timestamp: timestamppb.New(e.Content.Timestamp),
+			Metadata:  meta.GetStructValue(),
+			Fields:    fields.GetStructValue(),
+			// TODO this contains temporary values, since they are required and not available from the event at the moment
+			Input: &sc.Input{
+				Id:   "beats",
+				Name: "beats",
+				Type: "beats",
+			},
+			// TODO this contains temporary values, since they are required and not propagated at the moment
+			DataStream: &sc.DataStream{
+				// Id:        "none", // not generated at the moment
+				Type:      "shipper.output",
+				Dataset:   "generic",
+				Namespace: "default",
+			},
+		})
+	}
+
+	c.log.Debugf("all %d events converted to protobuf", len(events))
```

Comment:
Small detail on the text here, some events might not be converted due to errors, so I'd drop the "all":
```suggestion
	c.log.Debugf("%d events converted to protobuf", len(events))
```

Reply:
I don't think this change is worth to run the CI pipeline for 2 hours :D

Reply:
@belimawr but you're right, I also need to have a separate counter for non-dropped events or use the observer, I can fix it later, I've put it on my todo list.

Reply:
I've fixed the log message.